File: django_prosemirror/schema/nodes/ordered_list.py
"""Ordered list node definition."""

from prosemirror.model.schema import NodeSpec

from ..base import NodeDefinition


class OrderedListNode(NodeDefinition):
    """An ordered list."""

    @property
    def name(self) -> str:
        return "ordered_list"

    def to_dom(self, node) -> list:
        """Convert ordered list node to DOM representation."""
        attrs = self.class_mapping.apply_to_attrs(
            {"start": node.attrs["order"]} if "order" in node.attrs else {},
            "ordered_list",
        )
        return ["ol", attrs, 0] if attrs else ["ol", 0]

    def dom_matcher(self) -> list:
        """Return DOM parsing rules for ordered list."""
        return [
            {
                "tag": "ol",
                "getAttrs": lambda attrs: {
                    "start": attrs.get("start"),
                },
            }
        ]

    @property
    def spec(self) -> NodeSpec:
        return {
            "content": "list_item+",
            "group": "block",
            "parseDOM": self.dom_matcher(),
            "toDOM": self.to_dom,
        }

File: frontend/conf/schema/nodes/ordered_list.ts
import { type ClassMapping, NodeDefinition } from "@/conf/schema/abstract";
import { NodeType } from "@/conf/types";
import { DOMOutputSpec, Node, TagParseRule } from "prosemirror-model";

/**
 * Class that returns the spec of an ordered list node.
 * Prosemirror processes this node like ->
 * `<ol start={attrs.start}>{ content }</ol>`
 */
export class OrderedListNode extends NodeDefinition {
    /** Ordered list node. */
    constructor(classes: ClassMapping) {
        super(classes);
    }
    override name = NodeType.ORDERED_LIST;
    override attrs = { order: { default: 1, validate: "number" } };
    override content = "list_item+";
    override group = "block";
    override parseDOM: TagParseRule[] = [
        {
            tag: "ol",
            getAttrs(dom) {
                return {
                    order: dom.hasAttribute("start")
                        ? +dom.getAttribute("start")!
                        : 1,
                };
            },
        },
    ];
    override toDOM(node: Node): DOMOutputSpec {
        const attrs = this.classMapping.apply_to_attrs(
            { start: node?.attrs?.order ?? 1 },
            this.name,
        );
        return ["ol", attrs, 0];
    }
}
